crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -356,6 +356,139 @@ fn assert_component_access_compatibility(
     panic!("error[B0001]: Query<{query_type}, {filter_type}> in system {system_name} accesses component(s){accesses} in a way that conflicts with a previous system parameter. Consider using `Without<T>` to create disjoint Queries or merging conflicting Queries into a `ParamSet`. See: https://bevyengine.org/learn/errors/b0001");
 }
 
+// SAFETY: Relevant query ComponentId and ArchetypeComponentId access is applied to SystemMeta. If
+// this Query conflicts with any prior access, a panic will occur.
+unsafe impl<'a, D: QueryData + 'static, F: QueryFilter + 'static> SystemParam
+    for QuerySingle<'a, D, F>
+{
+    type State = QueryState<D, F>;
+    type Item<'w, 's> = QuerySingle<'w, D, F>;
+
+    fn init_state(world: &mut World, system_meta: &mut SystemMeta) -> Self::State {
+        Query::init_state(world, system_meta)
+    }
+
+    unsafe fn new_archetype(
+        state: &mut Self::State,
+        archetype: &Archetype,
+        system_meta: &mut SystemMeta,
+    ) {
+        // SAFETY: Delegate to existing `SystemParam` implementations.
+        unsafe { Query::new_archetype(state, archetype, system_meta) };
+    }
+
+    #[inline]
+    unsafe fn get_param<'w, 's>(
+        state: &'s mut Self::State,
+        system_meta: &SystemMeta,
+        world: UnsafeWorldCell<'w>,
+        change_tick: Tick,
+    ) -> Self::Item<'w, 's> {
+        state.validate_world(world.id());
+        // SAFETY: State ensures that the components it accesses are not accessible somewhere elsewhere.
+        let result =
+            unsafe { state.get_single_unchecked_manual(world, system_meta.last_run, change_tick) };
+        let single = result.unwrap();
+        QuerySingle {
+            item: single,
+            _filter: PhantomData,
+        }
+    }
+
+    #[inline]
+    unsafe fn validate_param(
+        state: &Self::State,
+        system_meta: &SystemMeta,
+        world: UnsafeWorldCell,
+    ) -> bool {
+        state.validate_world(world.id());
+        // SAFETY: State ensures that the components it accesses are not mutably accessible elsewhere
+        // and the query is read only.
+        let result = unsafe {
+            state.as_readonly().get_single_unchecked_manual(
+                world,
+                system_meta.last_run,
+                world.change_tick(),
```

Comment:
Are we sure this is the change tick that will be used for calling `get_param`, or an equivalent one for what matters to it? If so could you write a comment explaining why that's the case?

Reply:
Practically yes, executors work that way, but this is not a strict contract.
I think I can provide the exact `change_tick` that will be used for `run`, but I'll do that in another PR, is that fine?

Reply:
Practically yes, this happens to be the exact tick that is received by `get_param`, but there are not validation mechanisms/contract for that.

Currently only executors use `validate_param` and since the validation and running happens one after another, this is the correct change tick.
For `run_systems` and observers this also shouldn't be a problem.

While I can make `validate_param` accept a `change_tick`, the problem is that the tick used by `get_param` is generated inside `run_unsafe`, so we'd still need to uphold this "contract" of calling them back-to-back.

Reply:
I'm open for docs/mechanisms on making this better, I'm not sure how to go about it

Reply:
It won't always be exactly the same tick, because the multi-threaded executor will call `validate_param` and then spawn a task to run the system and call `get_param` at some later point.  Another system could start in between and increment the current tick.  

I think it will be equivalent, though.  Any components read by this param can't be written by another system in between the calls, so they can't have an `added` or `changed` tick in between the two `change_tick` values.

Reply:
Okay, I'll try to thoroughly document this in `validate_param`